bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -50,32 +49,37 @@ ClientsManager::ClientsManager(const std::set<NodeIdType>& proxyClients,
   std::copy(externalClients_.begin(), externalClients_.end(), std::ostream_iterator<NodeIdType>(oss, " "));
   oss << "internal clients: ";
   std::copy(internalClients_.begin(), internalClients_.end(), std::ostream_iterator<NodeIdType>(oss, " "));
-  LOG_DEBUG(CL_MNGR,
-            oss.str() << KVLOG(sizeOfReservedPage_, reservedPagesPerClient_, maxReplySize_, maxNumOfReqsPerClient_));
+  LOG_INFO(CL_MNGR,
+           oss.str() << KVLOG(sizeOfReservedPage(), reservedPagesPerClient_, maxReplySize_, maxNumOfReqsPerClient_));
 }
 
 uint32_t ClientsManager::reservedPagesPerClient(const uint32_t& sizeOfReservedPage, const uint32_t& maxReplySize) {
   uint32_t reservedPagesPerClient = maxReplySize / sizeOfReservedPage;
   if (maxReplySize % sizeOfReservedPage != 0) {
     reservedPagesPerClient++;
   }
+  reservedPagesPerClient++;  // for storing client public key
   return reservedPagesPerClient;
 }
 
-ClientsManager::~ClientsManager() { std::free(scratchPage_); }
-
 // Per client:
-// * calculate offset of reserved page start.
-// * load corresponding page from Reserved Pages to scratchPage.
-// * Fill its clientInfo.
-// * remove pending request if loaded reply is newer.
+// * load public key
+// * load page of the reply header
+// * fill clientInfo
+// * remove pending request if loaded reply is newer
 void ClientsManager::loadInfoFromReservedPages() {
   for (auto const& clientId : clientIds_) {
-    const uint32_t firstPageId = getFirstPageId(clientId);
+    if (loadReservedPage(getKeyPageId(clientId), sizeOfReservedPage(), scratchPage_.data())) {
+      auto& info = clientsInfo_[clientId];
+      std::istringstream iss(scratchPage_);
+      concord::serialize::Serializable::deserialize(iss, info.pubKey);
+      ConcordAssertGT(info.pubKey.first.length(), 0);
+      KeyExchangeManager::instance().loadClientPublicKey(info.pubKey.first, info.pubKey.second, clientId);
```

Comment:
What happens on the first start? 
The sigmanager initializes ```clientsPublicKeys_```   and the verifiers, from files and also this method calls ```SigManager::instance()->setClientPublicKey```

Reply:
At the first start this code won't execute since `loadReservedPage()` will return false

Reply:
and at consecutive starts, there is a dependency in the order i.e. first sigmanager then the ```loadClientPublicKey```, is it guaranteed to occur correctly?

Reply:
`SigManager::init()` is called from `ReplicaImp::ReplicaImp()` and `ClientsManager::loadReservedPages()` first time is called from `ReplicaImp::start()`